cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UsageHandler.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.gateway.handlers;
+
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.registry.UsageDataset;
+import co.cask.cdap.data2.registry.UsageDatasets;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.http.AbstractHttpHandler;
+import co.cask.http.HttpResponder;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import org.jboss.netty.handler.codec.http.HttpRequest;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+import java.util.Set;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * The {@link co.cask.http.HttpHandler} for handling REST calls to the usage registry.
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+@Singleton
+public class UsageHandler extends AbstractHttpHandler {
+
+  private final Supplier<UsageDataset> usageDataset;
+
+  @Inject
+  public UsageHandler(final DatasetFramework datasetFramework) {
+    this.usageDataset = Suppliers.memoize(new Supplier<UsageDataset>() {
+      @Override
+      public UsageDataset get() {
+        try {
+          return UsageDatasets.get(datasetFramework);
+        } catch (Exception e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    });
+  }
+
+  @GET
+  @Path("/namespaces/{namespace-id}/apps/{app-id}/datasets")
+  public void getAppDatasetUsage(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId,
+                                 @PathParam("app-id") String appId) {
+    Id.Application id = Id.Application.from(namespaceId, appId);
+    Set<? extends Id> ids = usageDataset.get().getDatasets(id);
```

Comment:
No, this was just to make it easier to copy paste, since these handler methods are the same except for the `UsageDataset` method it calls.